File: Multiple Linear Regression.R
# Load the dataset
# ToyotaCorolla.csv
data <- read.csv(file.choose())
View(data)

library(dplyr)
data <- data[,c("Price","Age_08_04","KM","HP","cc","Doors","Gears","Quarterly_Tax","Weight")]

#rename columns not for computer_data.csv
data <- data %>% rename(price = Price, age = Age_08_04, km = KM, hp = HP, cc= cc, doors = Doors, gears = Gears, qtax = Quarterly_Tax, weight = Weight) 

attach(data)

# price
# Normal distribution
qqnorm(price)
qqline(price)
# direction = +ve, strength = moderate, linearity = non linear

summary(data)
# Scatter plot | Plot relation ships between each X with Y
# age, price | qtax, price

plot(age, price) 
# direction = -ve, strength = moderate, linearity = non linear

plot(qtax, price)
# direction = +ve, strength = moderate, linearity = non linear

# Or make a combined plot | Scatter plot for all pairs of variables
pairs(data)   
plot(data)

cor(age, price)
# correlation matrix

cor(data) 
# weight-QuarterlyTax = high colinearity

# The Linear Model of interest | lm(Y ~ X)
model.data <- lm(price ~ age + km + hp + cc + doors + gears + qtax + weight, data = data)
summary(model.data)
# r-squared = 0.86, f value = significant, cc, doors insignificant > 0.05

# Scatter plot matrix with Correlations inserted in graph
library(GGally)
ggpairs(data)

# Partial Correlation matrix
library(corpcor)
cor(data)
cor2pcor(cor(data))

# Diagnostic Plots
library(car)

#Residual Plots, QQ-Plot, Std. Residuals vs Fitted, Cook's distance
plot(model.data)

#QQ plots of studentized residuals, helps identify outliers
qqPlot(model.data, id.n = 5)

# Deletion Diagnostics for identifying influential obseravations
# Index Plots of the influence measures
# A user friendly representation of the above

influenceIndexPlot(model.data, id.n = 3) 
influencePlot(model.data, id.n = 3) 
# 81 most influential observation

# Regression after deleting the observation
model.data1 <- lm(price ~ age + km + hp + cc + doors + gears + qtax + weight , data = data[-81, ])
summary(model.data1)
# r-squared = 0.86, f value = significant, doors insignificant > 0.05

# Variance Inflation Factors | VIF is > 10 => collinearity
vif(model.data1)   

# Regression model to check R^2 on Independent variales
VIF1 <- lm(price ~ age + km + hp + cc + doors + gears + qtax + weight )
VIF2 <- lm(age ~ price + km + hp + cc + doors + gears + qtax + weight )
VIF3 <- lm(km ~ age + price + hp + cc + doors + gears + qtax + weight )
VIF4 <- lm(hp ~ age + km + price + cc + doors + gears + qtax + weight )
VIF5 <- lm(cc ~ age + km + hp + price + doors + gears + qtax + weight )
VIF6 <- lm(doors ~ age + km + hp + cc + price + gears + qtax + weight )
VIF7 <- lm(gears ~ age + km + hp + cc + doors + price + qtax + weight )
VIF8 <- lm(qtax ~ age + km + hp + cc + doors + gears + price + weight )
VIF9 <- lm(weight ~ age + km + hp + cc + doors + gears + qtax + price )

summary(VIF1)
summary(VIF2)
summary(VIF3)
summary(VIF4)
summary(VIF5)
summary(VIF6)
summary(VIF7)
summary(VIF8)
summary(VIF9)

vif_rtable <- matrix(c(summary(VIF1)$r.squared,summary(VIF2)$r.squared,summary(VIF3)$r.squared,summary(VIF4)$r.squared,summary(VIF5)$r.squared,summary(VIF6)$r.squared,summary(VIF7)$r.squared,summary(VIF8)$r.squared,summary(VIF9)$r.squared),ncol=1,byrow=TRUE)
colnames(vif_rtable) <- c("R-squared")
rownames(vif_rtable) <- c("VIF1","VIF2","VIF3","VIF4","VIF5","VIF6","VIF7","VIF8","VIF9")
vif_rtable <- as.table(vif_rtable)
vif_rtable

# Added Variable Plots 
avPlots(model.data, id.n = 2, id.cex = 0.8, col = "red")

# Linear Model 
model.final <- lm(price ~ age + km + hp + cc + doors + gears + qtax + weight, data = data)
summary(model.final)
# r-squared = 0.86, f value = significant, cc, doors insignificant > 0.05

# Linear model without influential observation
model.final1 <- lm(price ~ age + km + hp + cc + gears + qtax + weight, data = data[-81, ])
summary(model.final1)
# r-squared = 0.86, f value = significant

rtable <- matrix(c(summary(model.data)$r.squared,summary(model.data1)$r.squared,summary(model.final)$r.squared,summary(model.final1)$r.squared),ncol=1,byrow=TRUE)
colnames(rtable) <- c("R-squared")
rownames(rtable) <- c("model1","model2","model3","model4")
rtable <- as.table(rtable)
rtable

# Added Variable Plots
avPlots(model.final1, id.n = 2, id.cex = 0.8, col = "red")

# Variance Influence Plot
vif(model.final1)

# Evaluation Model Assumptions
plot(model.final1)
plot(model.final1$fitted.values, model.final1$residuals)

qqnorm(model.final1$residuals)
qqline(model.final1$residuals)

library(leaps)
lm_best <- regsubsets(price ~ ., data = data, nvmax = 15)
summary(lm_best)
summary(lm_best)$adjr2
which.max(summary(lm_best)$adjr2)
coef(lm_best, 3)

lm_forward <- regsubsets(price ~ ., data = data, nvmax = 15, method = "forward")
summary(lm_forward)

# Data Partitioning
n <- nrow(data)
n1 <- n * 0.8
n2 <- n - n1
train <- sample(1:n, n1)
test <- data[-train, ]

# Model Training
# price ~ age + km + hp + cc + gears + qtax + weight
model <- lm(price ~ age + km + hp + cc + gears + qtax + weight, data[train, ])
summary(model)

pred <- predict(model, newdata = test)
actual <- test$price
error <- actual - pred

test.rmse <- sqrt(mean(error**2))
test.rmse

train.rmse <- sqrt(mean(model$residuals**2))
train.rmse
# train = 1305.96, test = 1471.66

File: Multiple Linear Regression.py
import pandas as pd
import numpy as np

# loading the data
data = pd.read_csv("ToyotaCorolla.csv")
data.describe()

data_new = data.filter(['Price','Age_08_04','KM','HP','cc','Doors','Gears','Quarterly_Tax','Weight'], axis = 1)
data_new.columns = "Price","Age","KM","HP","cc","Doors","Gears","QuarterlyTax","Weight"
data_new.describe()

#Graphical Representation
import matplotlib.pyplot as plt

# Price
data_new.Price.plot.bar()
plt.hist(data_new.Price) #histogram
plt.boxplot(data_new.Price) #boxplot
# 1 peak, many outliers

# HP
data_new.HP.plot.bar()
plt.hist(data_new.HP) #histogram
plt.boxplot(data_new.HP) #boxplot
# 1 peak, some outliers

# Jointplot
import seaborn as sns

# Price, Age
sns.jointplot(x=data_new['Price'], y=data_new['Age'])
# direction = -ve, strength = moderate, linearity = non linear

# Countplot - 1 figure, 16 = height, 10 = width
plt.figure(1, figsize=(16, 10))
sns.countplot(data_new['Age'])

# Q-Q Plot
from scipy import stats
import pylab
stats.probplot(data_new.Price, dist = "norm", plot = pylab)
plt.show()
# non linear

# Scatter plot between the variables along with histograms
import seaborn as sns
sns.pairplot(data_new.iloc[:, :])

# Correlation matrix 
data_new.corr()
# weight-QuarterlyTax = high colinearity

# preparing model considering all the variables 
import statsmodels.formula.api as smf # for regression model

ml1 = smf.ols('Price ~ Age + KM + HP + cc + Doors + Gears + QuarterlyTax + Weight', data = data_new).fit() 
ml1.summary()
# r-squared = 0.86, f value = significant, cc, doors insignificant > 0.05

# Checking whether data has any influential values 
# Influence Index Plots
import statsmodels.api as sm

sm.graphics.influence_plot(ml1)
# 80 most influential observation

# Studentized Residuals = Residual/standard deviation of residuals
#drop influential observation
data_new = data_new.drop(data_new.index[[80]])

# Preparing model                  
ml_new = smf.ols('Price ~ Age + KM + HP + cc + Doors + Gears + QuarterlyTax + Weight', data = data_new).fit()    
ml_new.summary()
# r-squared = 0.86, f value = significant, doors insignificant > 0.05

# Check for Colinearity to decide to remove a variable using VIF
# Assumption: VIF > 10 = colinearity
# calculating VIF's values of independent variables
#figure which variable has highest vif value
rsq_1 = smf.ols('Price ~ Age + KM + HP + cc + Doors + Gears + QuarterlyTax + Weight', data = data_new).fit().rsquared  
vif_1 = 1/(1 - rsq_1) 

rsq_2 = smf.ols('KM ~ Age + Price + HP + cc + Doors + Gears + QuarterlyTax + Weight', data = data_new).fit().rsquared  
vif_2 = 1/(1 - rsq_2)

rsq_3 = smf.ols('HP ~ Age + KM + Price + cc + Doors + Gears + QuarterlyTax + Weight', data = data_new).fit().rsquared  
vif_3 = 1/(1 - rsq_3) 

rsq_4 = smf.ols('cc ~ Age + KM + HP + Price + Doors + Gears + QuarterlyTax + Weight', data = data_new).fit().rsquared  
vif_4 = 1/(1 - rsq_4) 

rsq_5 = smf.ols('Doors ~ Age + KM + HP + cc + Price + Gears + QuarterlyTax + Weight', data = data_new).fit().rsquared  
vif_5 = 1/(1 - rsq_5) 

rsq_6 = smf.ols('Gears ~ Age + KM + HP + cc + Doors + Price + QuarterlyTax + Weight', data = data_new).fit().rsquared  
vif_6 = 1/(1 - rsq_6) 

rsq_7 = smf.ols('QuarterlyTax ~ Age + KM + HP + cc + Doors + Gears + Price + Weight', data = data_new).fit().rsquared  
vif_7 = 1/(1 - rsq_7) 

rsq_8 = smf.ols('Weight ~ Age + KM + HP + cc + Doors + Gears + QuarterlyTax + Price', data = data_new).fit().rsquared  
vif_8 = 1/(1 - rsq_8) 

rsq_9 = smf.ols('Age ~ Price + KM + HP + cc + Doors + Gears + QuarterlyTax + Weight', data = data_new).fit().rsquared  
vif_9 = 1/(1 - rsq_9) 

# Storing vif values in a data frame
d1 = {'Variables':['Price', 'KM', 'HP', 'cc', 'Doors', 'Gears', 'QuarterlyTax', 'Weight', 'Age'], 'VIF':[vif_1, vif_2, vif_3, vif_4, vif_5, vif_6, vif_7, vif_8, vif_9]}
Vif_frame = pd.DataFrame(d1)  
Vif_frame
# Price highest vif
# we are going to drop the column having highest vif from the prediction model

# Final model
final_ml = smf.ols('QuarterlyTax ~ KM + HP + cc + Doors + Gears + Age + Weight', data = data_new).fit()
final_ml.summary() 
# r-squared = 0.65, f value = significant

# Prediction
pred = final_ml.predict(data_new)

# Q-Q plot
res = final_ml.resid

# Q-Q plot
stats.probplot(res, dist = "norm", plot = pylab)
plt.show()
# not normally distributed

# Residuals vs Fitted plot
sns.residplot(x = pred, y = data_new.Price, lowess = True)
plt.xlabel('Fitted')
plt.ylabel('Residual')
plt.title('Fitted vs Residual')
plt.show()

sm.graphics.influence_plot(final_ml)
# 221 most influential observation

# Splitting the data into train and test data 
from sklearn.model_selection import train_test_split
data_train, data_test = train_test_split(data_new, test_size = 0.2) 

# preparing the model on train data 
# QuarterlyTax ~ KM + HP + cc + Doors + Gears + Age + Weight
model_train = smf.ols("QuarterlyTax ~ KM + HP + cc + Doors + Gears + Age + Weight", data = data_train).fit()

# prediction on test data set 
test_pred = model_train.predict(data_test)
# test residual values 
test_resid = test_pred - data_test.QuarterlyTax
# RMSE value for test data 
test_rmse = np.sqrt(np.mean(test_resid * test_resid))
test_rmse

# train_data prediction
train_pred = model_train.predict(data_train)

# train residual values 
train_resid  = train_pred - data_train.QuarterlyTax
# RMSE value for train data 
train_rmse = np.sqrt(np.mean(train_resid * train_resid))
train_rmse
# training = 23.43,testing = 27.93